src/Clapi/Valuespace.hs

```diff
@@ -288,24 +289,40 @@ eitherErrorMap =
     Map.partition isLeft
 
 vsValidate :: Valuespace OwnerUnvalidated -> MonadErrorMap (Valuespace Validated)
-vsValidate =
-    rectifyTypes >=> validateChildren >=> validateData >=> markValid
+vsValidate vs =
+    rectifyTypes vs (view (unvalidated . etas) vs) >>= validateChildren >>= validateData >>= markValid
   where
     markValid = return . set unvalidated ()
 
-rectifyTypes ::
-    (HasUvtt v TaintTracker) => Valuespace v -> MonadErrorMap (Valuespace v)
-rectifyTypes vs =
-  do
-    unvalidatedsTypes <-
-        eitherErrorMap . Map.mapWithKey (\np _ -> getType vs np) .
-        view (unvalidated . uvtt) $ vs
-    newDefs <-
-        eitherErrorMap . Map.mapWithKey toDef $ toMetaTypes unvalidatedsTypes
-    return . over defs (Map.union newDefs) $ vs
+rectifyTypes :: (HasUvtt v TaintTracker) => Valuespace v -> Set.Set NodePath -> MonadErrorMap (Valuespace v)
+rectifyTypes vs explicitPaths = populateDefsFor (Map.keysSet (view (unvalidated . uvtt) vs)) vs
   where
-    toMetaTypes = mapFilterJust . fmap categoriseMetaTypePath
-    toDef np mt = getNode np vs >>= validateTypeNode mt
+    explicitlyAssigned = Map.restrictKeys (fst $ view types vs) explicitPaths
+    populateDefsFor :: Set.Set NodePath -> Valuespace v -> MonadErrorMap (Valuespace v)
+    populateDefsFor deflessNodes vs = if null deflessNodes then (mempty, vs) else let
+        np = head $ Set.toList deflessNodes
+        cftad = getOrBuildDef deflessNodes np
```

Comment:
I have no clue what `cftad` is, and you only use it once below anyway, so you could just make line 306
```haskell
vs' <- case getOrBuildDef deflessNodes np of
    Left err -> blah
    Right (tp, md) -> bleh
```